### PR TITLE
fix: crypto is not defined

### LINKED ISSUE #135: crypto is not defined
### Is this a regression?

Yes

### Description

I'm running `@atala/prism-wallet-sdk@3.2.0` in Node, and I got this error?

### Please provide the exception or error you saw

```bash
ReferenceError: crypto is not defined
  at protoOf.kv (/home/projects/ssb-atala-prism/node_modules/@atala/apollo/Apollo.js:1734:26)
  at protoOf.jv (/home/projects/ssb-atala-prism/node_modules/@atala/apollo/Apollo.js:1272:45)
  at protoOf.createRandomMnemonics (/home/projects/ssb-atala-prism/node_modules/@atala/apollo/Apollo.js:1577:39)
  at Apollo.createRandomSeed (/home/projects/ssb-atala-prism/node_modules/@atala/prism-wallet-sdk/build/node-cjs/index.cjs:1637:36)
  at createAgent (/home/projects/ssb-atala-prism/create-agent.js:22:23)
```

### Please provide the environment you discovered this bug in

_No response_

### Anything else?

Line causing the issue is a piece of code

```js
    return toByteArray_1(crypto.getRandomValues(arr).buffer);
```

### PR DESCRIPTION
**Pull request by hamada147**
# Overview
Fix crypto is not defined

Fixes #135 

### My PR contains...
* [ ] No code changes (changes to documentation, CI, metadata, etc.)
* [x] Bug fixes (non-breaking change which fixes an issue)
* [ ] Improvements (misc. changes to existing features)
* [ ] Features (non-breaking change which adds functionality)

### My changes...
* [ ] are breaking changes
* [x] are not breaking changes
* [ ] If yes to above: I have updated the documentation accordingly

### Documentation
* [x] My changes do not require a change to the project documentation
* [ ] My changes require a change to the project documentation
* [ ] If yes to above: I have updated the documentation accordingly

### Tests
* [ ] My changes can not or do not need to be tested
* [x] My changes can and should be tested by unit and/or integration tests
* [x] If yes to above: I have added tests to cover my changes
* [x] If yes to above: I have taken care to cover edge cases in my tests

### CODE REVIEW

**Comment by atala-dev** [comment]
### Code Coverage
|File|Coverage [100.00%]|:white_check_mark:|
|:-|:-:|:-:|
|[apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/securerandom/SecureRandom.kt](https://github.com/input-output-hk/atala-prism-apollo/blob/679cb77e51564c2164a35b1e48ea83d4686ceac3/apollo%2Fsrc%2FjsMain%2Fkotlin%2Fio%2Fiohk%2Fatala%2Fprism%2Fapollo%2Fsecurerandom%2FSecureRandom.kt)|100.00%|:white_check_mark:|

|Total Project Coverage|91.66%|:white_check_mark:|
|:-|:-:|:-:|

File: apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/utils/HelperMethods.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * A variable that represents the global object in JavaScript.
+ *
+ * The `global` variable is initialized lazily using the `lazy` delegate.
+ * It is assigned a dynamic value that depends on the current execution environment.
+ *
+ * In a Node.js environment, the `global` variable refers to the global object.
+ * In a browser environment, the `global` variable refers to the `window` object.
+ */
+val global: dynamic by lazy {
+    js("((typeof global !== 'undefined') ? global : self)")
```
Comment (curtis-h) [inline apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/utils/HelperMethods.kt:23]: self is not contextually guaranteed, should be `window` for browsers
```suggestion
    js("((typeof global !== 'undefined') ? global : window)")
```

File: apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/utils/HelperMethods.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    when (runtime) {
+        null -> {
+            js("(typeof global !== 'undefined' && ({}).toString.call(global) == '[object global]')") as Boolean
+        }
+        "node" -> true
+        else -> false
+    }
+}
```
Comment (curtis-h) [inline apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/utils/HelperMethods.kt:47]: building on comment above, adding crypto alias:
```suggestion
}

val crypto: dynamic by lazy {
  return if (!global.crypto && isNode) {
     _require("crypto")
   } else {
     global.crypto
   }
}
```

File: apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/securerandom/SecureRandom.kt [hunk 3]
```diff
@@ -19,13 +20,21 @@ actual class SecureRandom actual constructor(
      */
     override fun nextBytes(size: Int): ByteArray {
         val arr = Uint8Array(size)
-        return crypto.getRandomValues(arr).buffer.toByteArray()
+        return if (isNode) {
+            _require("crypto").randomFillSync(arr)
+        } else {
+            global.crypto.getRandomValues(arr)
+        }
```
[the remark this answers is not in the document]
Comment (hamada147) [reply apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/securerandom/SecureRandom.kt:27]: Because:
- The `crypto` module is not the same for the `Node.js` and `Browser` environments. Both have different usage interfaces.
- If the `Node.js` module is used in `Browser`, it fails with the module not found, and the same happens when reversed.

That's why I implemented them like that.

For reference, check #135

File: apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/securerandom/SecureRandom.kt [hunk 4]
```diff
@@ -19,13 +20,21 @@ actual class SecureRandom actual constructor(
...
 
     actual companion object : SecureRandomStaticInterface {
         override fun generateSeed(numBytes: Int): ByteArray {
             val arr = Uint8Array(numBytes)
-            return crypto.getRandomValues(arr).buffer.toByteArray()
+            return if (isNode) {
+                _require("crypto").randomFillSync(arr)
+            } else {
+                global.crypto.getRandomValues(arr)
+            }
```
[the remark this answers is not in the document]
Comment (hamada147) [reply apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/securerandom/SecureRandom.kt:37]: We can't because they are very different. The `crypto` module is different for the `Node.js` and `Browser`. Both have other usage interfaces.

If we want/need to implement it like you are suggesting. We would create our interface and class that implements that interface, and in each method internally, we would have the if condition based on the environment.

File: apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/utils/HelperMethods.kt (hunk 1 above)
Comment (hamada147) [reply apollo/src/jsMain/kotlin/io/iohk/atala/prism/apollo/utils/HelperMethods.kt:23]: But, In a Node.js environment, the `global` variable refers to the global object; in a browser environment, the `global` variable refers to the `window` object. That's why it should be kept as `self`. This way, in each environment, I would get the correct `global` variable.

**Review by curtis-h — APPROVED** [review]
